### PR TITLE
[client] Local client builder

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR implements `ray.client("local").connect()`, as well as making `ray.client().connect()` much smarter, by making it either create, or autodetect a cluster running on the same node.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
Suggestion: this PR is based off of #15934, so merge that first to make this PR much smaller.

**Comment by wuisawesome** [comment]
cc @pcmoritz @ericl @richardliaw

File: python/ray/scripts/scripts.py [hunk 1]
```diff
@@ -800,6 +802,11 @@ def stop(force, verbose, log_style, log_color):
             cli_logger.warning("Try running the command again, or use `{}`.",
                                cf.bold("--force"))
 
+    try:
+        os.remove("/tmp/ray/current_cluster")
+    except OSError:
+        # This just means the file doesn't exist.
```
Comment (AmeerHajAli) [inline python/ray/scripts/scripts.py:808]: what do we do in that case?

File: python/ray/client_builder.py [hunk 2]
```diff
@@ -77,7 +81,16 @@ def _split_address(address: str) -> Tuple[str, str]:
...
+    if address == "local":
+        return _LocalClientBuilder(None)
+    if address is None:
+        try:
+            with open("/tmp/ray/current_cluster", "r") as f:
+                address = f.read()
+                print(address)
+        except FileNotFoundError:
+            # `address` won't be set and we'll create a new cluster.
+            pass
```
Comment (AmeerHajAli) [inline python/ray/client_builder.py:93]: why not to check if the file exists and then to read it?  Also, why are we printing the address here?

File: python/ray/scripts/scripts.py [hunk 3]
```diff
@@ -800,6 +802,11 @@ def stop(force, verbose, log_style, log_color):
             cli_logger.warning("Try running the command again, or use `{}`.",
                                cf.bold("--force"))
 
+    try:
+        os.remove("/tmp/ray/current_cluster")
```
Comment (AmeerHajAli) [inline python/ray/scripts/scripts.py:806]: if file exists remove?

File: python/ray/tests/test_client_builder.py [hunk 4]
```diff
@@ -99,3 +100,69 @@ def test_connect_to_cluster(ray_start_regular_shared):
     assert client_info.protocol_version == protocol_version
 
     server.stop(0)
+
+
+def test_local_clusters():
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_builder.py:105]: Can you please add description of what this is testing and how?

File: python/ray/tests/test_client_builder.py [hunk 5]
```diff
@@ -99,3 +100,69 @@ def test_connect_to_cluster(ray_start_regular_shared):
...
+a = Foo.options(name="abc", lifetime="detached").remote()
+ray.get(a.ping.remote())
+
+if {blocking}:
+    import time
+    while True:
+        time.sleep(30)
+
+"""
+    # no_address_script = driver_template.format(address="")
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_builder.py:124]: ?

File: python/ray/scripts/scripts.py (hunk 3 above)
Comment (wuisawesome) [reply python/ray/scripts/scripts.py:806]: I think this is the properly pythonic way to do it. https://docs.python.org/3/glossary.html#term-eafp

File: python/ray/tests/test_client_builder.py [hunk 6]
```diff
@@ -99,3 +100,69 @@ def test_connect_to_cluster(ray_start_regular_shared):
...
+
+@ray.remote
+class Foo:
+    def ping(self):
+        return "pong"
+
+a = Foo.options(name="abc", lifetime="detached").remote()
+ray.get(a.ping.remote())
+
+if {blocking}:
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_builder.py:118]: this does not seem to be used

**Review by AmeerHajAli — APPROVED** [review]
LGTM, left a few comments

File: python/ray/tests/test_client_builder.py (hunk 6 above)
Comment (wuisawesome) [reply python/ray/tests/test_client_builder.py:118]: god point...

File: python/ray/scripts/scripts.py (hunk 3 above)
Comment (AmeerHajAli) [reply python/ray/scripts/scripts.py:806]: I am not sure this is what it means though, you also have code with if exists and avoids an exceptional case, but won't push too hard.

**Comment by wuisawesome** [comment]
Flaky build is flaky, everything else is green. Merging

**Comment by AmeerHajAli** [comment]
@wuisawesome,  I marked this a release blocker, can you please add this to the spreadsheet?
cc @DmitriGekhtman .

@ijrsvt, can we please add some experimental documentation for this ray API?

**Pull request merged**